source/MaterialXTest/MaterialXFormat/GraphIO.cpp

```diff
@@ -0,0 +1,136 @@
+    for (const std::string& libraryName : libraryNames)
+    {
+        mx::FilePath libPath = outputPath / libraryName;
+        libPath.createDirectory();
+    }
+
+    // Create log file.
+    const mx::FilePath logPath("graphio_test.txt");
+    std::ofstream logFile;
+    logFile.open(logPath);
+
+    bool failedGeneration = false;
+
+    mx::GraphIORegistryPtr graphIORegistry = mx::GraphIORegistry::create();
+    graphIORegistry->addGraphIO(mx::MermaidGraphIO::create());
+    graphIORegistry->addGraphIO(mx::DotGraphIO::create());
+
+    mx::StringVec extensions = { "md", "dot" };
+
+    for (const mx::NodeDefPtr& nodedef : stdlib->getNodeDefs())
+    {
+        std::string nodeName = nodedef->getName();
+        std::string nodeNode = nodedef->getNodeString();
+        if (nodeName.size() > 3 && nodeName.substr(0, 3) == "ND_")
+        {
+            nodeName = nodeName.substr(3);
+        }
+
+        mx::InterfaceElementPtr interface = nodedef->getImplementation();
+        mx::NodeGraphPtr nodegraph = interface ? interface->asA<mx::NodeGraph>() : nullptr;
+        if (!nodegraph)
+        {
+            logFile << "Skip generating for node with graph representation '" << nodeName << "'" << std::endl;
+            continue;
+        }
+
+        try
+        {
+            std::vector<mx::OutputPtr> roots;
+            std::vector<bool> writeCategoriesList = { true, false };
+
+            for (const std::string& extension : extensions)
+            {
+                mx::GraphIOGenOptions graphOptions;
+                graphOptions.setWriteSubgraphs(false);
+                graphOptions.setOrientation(mx::GraphIOGenOptions::Orientation::LEFT_RIGHT);
+
+                for (auto writeCategories : writeCategoriesList)
+                {
+                    graphOptions.setWriteCategories(writeCategories);
+                    std::string graphString = graphIORegistry->write(extension, nodegraph, roots, graphOptions);
+                    if (!graphString.empty())
+                    {
+                        logFile << "Wrote " + (!writeCategories ? " instanced " : mx::EMPTY_STRING) + "graph for node '" << nodeName << "'" << std::endl;
+
+                        mx::FilePath libPath = outputPath;
+                        for (const std::string& libraryName : libraryNames)
+                        {
+                            const std::string uri = nodegraph->getSourceUri();
+                            if (std::string::npos != uri.find(libraryName))
+                            {
+                                libPath = outputPath / libraryName;
+                                break;
+                            }
+                        }
+
+                        const std::string filename = nodeName + (!writeCategories ? "_instance" : mx::EMPTY_STRING) + "." + extension;
+                        const std::string filepath = (libPath / filename).asString();
+                        std::ofstream file;
```

Comment:
Generates a directory of all definitions Markdown and dot graphs. This is not saved out anywhere for later usage but could be.